File: iteration3/src/model/transcript.py
from typing import List

from model.course import Course
from model.takencourse import TakenCourse

from utils.utils import gpa_letter_to_grade


class Transcript(object):

    def _init_(self):
        self.__semester: int = 0
        self.__semesterGPA: float = 0
        self.__totalCredit: float = 0
        self.__id: int = 0
        self.__semesterCourses: List[TakenCourse] = []

    def calculate(self):
        self.calculateTotalCredit()
        self.calculateSemesterGPA()

    def calculateSemesterGPA(self):
        creditCount = 0
        sum = 0.0
        for c in self.__semesterCourses:
            creditCount += c.__credit
            sum += c.__credit * gpa_letter_to_grade.get(str(c.__letterGrade), 0.0)

        self.__semesterGPA = sum/creditCount

    def calculateTotalCredit(self):
        sum = 0
        for course in self.__semesterCourses:
            sum += course.getCredit()
        self.__totalCredit = sum

    def getCourses(self):
        return self.__semesterCourses

    def getSemester(self):
        return self.__semester

    def setSemester(self, semester: int):
        self.__semester = semester

    def getSemesterGPA(self):
        return self.__semesterGPA

    def getTotalCredit(self):
        return self.__totalCredit

    def getSemesterCourses(self):
        return self.__semesterCourses

    def addCourse(self, tcs: List [Course]):
        if self.__semesterCourses is None:
            self.__semesterCourses = []
        self.__semesterCourses.append(tcs)

    def add_Course(self, tcs: List [Course]):
        if self.__semesterCourses is None:
            self.__semesterCourses = []
        self.__semesterCourses.extend(tcs)

File: iteration1/code/src/coursetracking/models/Transcript.java
package coursetracking.models;

import java.util.ArrayList;

import com.google.gson.annotations.SerializedName;

import coursetracking.utils.Utils;

public class Transcript implements Comparable<Transcript> {
    private int id;

    @SerializedName("semester")
    private int semester;

    private ArrayList<TakenCourse> semesterCourses;

	private float semesterGPA;
    private int totalCredit;

    public void calculate() {
        calculateTotalCredit();
        calculateSemesterGPA();
    }

    public void calculateSemesterGPA() {
        int creditCount = 0;
        Utils utils = Utils.getInstance();
        float sum = 0.0f;
        for (TakenCourse c : semesterCourses) {
            creditCount += c.getCredit();
            sum += c.getCredit() * utils.getGPAOfLetterGrade(c.getLetterGrade());
        }

        this.semesterGPA = sum/creditCount;
    }

    public void calculateTotalCredit() {
        int sum = 0;
        for (TakenCourse course: semesterCourses) {
            sum += course.getCredit();
        }
        totalCredit = sum;
    }

    public ArrayList<TakenCourse> getCourses() {
        return semesterCourses;
    }

    public int getSemester() {
        return semester;
    }

    public void setSemester(int semester) {
        this.semester = semester;
    }

    public float getSemesterGPA() {
        return semesterGPA;
    }

    public int getTotalCredit() {
        return totalCredit;
    }
    
    public ArrayList<TakenCourse> getSemesterCourses() {
		return semesterCourses;
	}

    @Override
    public int compareTo(Transcript o) {
        return o.semester;
    }   

    public void addCourse(TakenCourse tc){
        if (this.semesterCourses == null) this.semesterCourses = new ArrayList<>();
        this.semesterCourses.add(tc);
    }

    public void addCourse(ArrayList<TakenCourse> tcs) {
        if (this.semesterCourses == null) this.semesterCourses = new ArrayList<>();
        this.semesterCourses.addAll(tcs);
    }
}
